packages/protocol/contracts-0.8/common/EpochManagerInitializer.sol

```diff
@@ -0,0 +1,62 @@
+// SPDX-License-Identifier: UNLICENSED
+pragma solidity >=0.8.7 <0.8.20;
+
+import "../common/UsingRegistry.sol";
+import "../common/UsingPrecompiles.sol";
+
+import "../../contracts/common/Initializable.sol";
+import "../../contracts/common/interfaces/ICeloVersionedContract.sol";
+import "../../contracts/governance/interfaces/IEpochRewards.sol";
+
+contract EpochManagerInitializer is Initializable, UsingPrecompiles, UsingRegistry {
+  /**
+   * @notice Sets initialized == true on implementation contracts
+   * @param test Set to true to skip implementation initialization
+   */
+  constructor(bool test) public Initializable(test) {}
+
+  /**
+   * @notice Used in place of the constructor to allow the contract to be upgradable via proxy.
+   * @param registryAddress The address of the registry core smart contract.
+   */
+  function initialize(address registryAddress) external initializer {
+    _transferOwnership(msg.sender);
+    setRegistry(registryAddress);
+  }
+
+  /**
+   * @notice initializes the epochManager contract during L2 transition.
+   */
+  function initEpochManager() external onlyOwner {
+    uint256 currentEpoch = getEpochNumber();
+
+    uint256 numberElectedValidators = numberValidatorsInCurrentSet();
+
+    address[] memory electedValidatorAddresses = new address[](numberElectedValidators);
+
+    for (uint256 i = 0; i < numberElectedValidators; i++) {
+      address validatorAddress = validatorSignerAddressFromCurrentSet(i);
+      electedValidatorAddresses[i] = validatorAddress;
+    }
+    getEpochManager().initializeSystem(
+      currentEpoch,
+      _getFirstBlockOfEpoch(currentEpoch),
+      electedValidatorAddresses
+    );
+  }
+
+  function getFirstBlockOfEpoch(uint256 currentEpoch) external view returns (uint256) {
+    return _getFirstBlockOfEpoch(currentEpoch);
+  }
+
+  function _getFirstBlockOfEpoch(uint256 currentEpoch) internal view returns (uint256) {
```

Comment:
this is quite suboptimal.

Either do a binary search to check if an block is what epoch, or check `pos.go` to implement this functions

There's a solidity version of epochSize and with that you can compute epoch for block X pretty easy

From pos.go
```
// GetEpochNumber retrieves the epoch number given the block number.
// Epoch 0 is a special block that only contains the genesis block (block 0), epoch 1
// starts at block 1
func GetEpochNumber(number uint64, epochSize uint64) uint64 {
	if IsLastBlockOfEpoch(number, epochSize) {
		return number / epochSize
	} else {
		return number/epochSize + 1
	}
}

// MustGetEpochFirstBlockGivenBlockNumber is a variant of GetEpochFirstBlockGivenBlockNumber
// that panics if called for epoch 0 (genesis)
func MustGetEpochFirstBlockGivenBlockNumber(blockNumber uint64, epochSize uint64) uint64 {
	firstBlock, err := GetEpochFirstBlockGivenBlockNumber(blockNumber, epochSize)
	if err != nil {
		panic(err)
	}
	return firstBlock
}

// GetEpochFirstBlockGivenBlockNumber retrieves first block of a given block's epoch
// Fails when try to obtain first block of epoch 0 (genesis)
func GetEpochFirstBlockGivenBlockNumber(blockNumber uint64, epochSize uint64) (uint64, error) {
	epochNumber := GetEpochNumber(blockNumber, epochSize)
	return GetEpochFirstBlockNumber(epochNumber, epochSize)
}

// GetEpochFirstBlockNumber retrieves first block of epoch.
func GetEpochFirstBlockNumber(epochNumber uint64, epochSize uint64) (uint64, error) {
	// Epoch 0 is just the genesis block, it doesn't have a first block (only last)
	if epochNumber == 0 {
		return 0, errors.New("No first block for epoch 0")
	}

	return ((epochNumber - 1) * epochSize) + 1, nil
}
```